### DISCUSSION TITLE
Create Calculated Fields in Tableau using Python

### DISCUSSION THREAD

**Discussion by jsnyder2k** [tableau/TabPy#607] [Q&A]
I might be asking too much of Tabpy, but is it possible to create calculated fields in Tableau using Python? I'm using Tableau desktop. I have hundreds of calculations I need to create in Tableau and I usually do it with a macro key program to click through the various menus in Tableau and copy data from Excel. I want something that is more reliable and I thought Tabpy can do it by reading my Excel sheet with my calculations and building them in Tableau. I got this far:

```
import pandas as pd
from tableaudocumentapi import Workbook
import subprocess
import time
from tabpy.tabpy_tools.client import Client

# Start the TabPy server
subprocess.Popen(["tabpy"])

# Wait for the server to start (adjust the sleep duration if needed)
time.sleep(5)

# Connect to the TabPy server
client = Client('http://localhost:9004/')

# Path to the Excel file
excel_file_path = 'C:/Users/Documents/PythonTableau/Python.xlsx'

# Tableau workbook path
workbook_path = 'C:/Users/Documents/PythonTableau/PythonTab.twb'

# Read the Excel file into a DataFrame
df = pd.read_excel(excel_file_path)

# Open the workbook
wb = Workbook(workbook_path)

# Find the worksheet with the matching name
worksheet_name = 'Sheet 1'
worksheet = None
for ws in wb.worksheets:
    if ws.name == worksheet_name:
        worksheet = ws
        break

if worksheet is None:
    print(f"Worksheet '{worksheet_name}' not found in the workbook.")
    exit()

# Iterate over the rows of the DataFrame
for _, row in df.iterrows():
    calculated_field_name = row['Calculated Field']
    calculated_field_formula = row['Calculation']

    # Create the calculated field
    calc_field = worksheet.create_calculated_field(calculated_field_name, calculated_field_formula)

    print(f"Calculated field '{calculated_field_name}' created in Tableau workbook.")

# Save the workbook with the updated calculated fields
wb.save(workbook_path)

print("All calculated fields created in Tableau workbook.")
```

I've tried lots of ways to make it work, but I think I'm stuck on finding the sheet name for some reason. I've tried just pointing to the workbook, but I'm not even sure if it's possible to do this.

the specific error I'm getting on this code is:

```
Traceback (most recent call last):
  File "C:\Users\Documents\PythonTableau\TableauCalculation.py", line 32, in <module>
    if ws.name == worksheet_name:
AttributeError: 'str' object has no attribute 'name'
```

Any help on the possibility of this would be most helpful.

Thank you!

**Comment by AmirMK** [comment]
You can use table extension. Table extension works based on Tabpy and enable you to run your python code at datasource level and generate a full table as output. The best part is the input of your python code can be a table generated by tableau data model.

**Reply by AmirMK** [reply]
Here is detailed video: https://m.youtube.com/watch?v=ATR8WVp9ovo
